### DISCUSSION TITLE
ACE_Refcounted_Auto_Ptr_Rep pointer assignment to "NULL" or "0"

### DISCUSSION THREAD

**Discussion by ashaikh111** [DOCGroup/ACE_TAO#1847] [Q&A]
ACE version 6.3.4

Hi,
I am new to ACE,
In the following code block, why "rep" is not assigned with "NULL" or "0" after deleting it.
is it bug, that it is not assigned with "NULL" or "0" ?

```c++
template <class X, class ACE_LOCK> inline void
ACE_Refcounted_Auto_Ptr_Rep<X, ACE_LOCK>::detach (ACE_Refcounted_Auto_Ptr_Rep<X, ACE_LOCK>*& rep)
{
if (rep == 0)
return;

if (rep->ref_count_-- == 0) {
delete rep;
}
```

**Comment by iguessthislldo** [comment]
I'm not familiar with this class at all, so I might be wrong about this, but here are the places I could find that use `detach`:

https://github.com/DOCGroup/ACE_TAO/blob/1522d0f3572872020a2eb8167d477c0efa965c50/ACE/ace/Refcounted_Auto_Ptr.inl#L143-L144

https://github.com/DOCGroup/ACE_TAO/blob/1522d0f3572872020a2eb8167d477c0efa965c50/ACE/ace/Refcounted_Auto_Ptr.inl#L178-L179

In these two uses it sets to null after `detach` like it should.

https://github.com/DOCGroup/ACE_TAO/blob/1522d0f3572872020a2eb8167d477c0efa965c50/ACE/ace/Refcounted_Auto_Ptr.inl#L156-L160

https://github.com/DOCGroup/ACE_TAO/blob/1522d0f3572872020a2eb8167d477c0efa965c50/ACE/ace/Refcounted_Auto_Ptr.inl#L171-L174

In these two cases it's replacing the pointer so it doesn't matter what it does with the old pointer value. Assuming it's not used again.

https://github.com/DOCGroup/ACE_TAO/blob/1522d0f3572872020a2eb8167d477c0efa965c50/ACE/ace/Refcounted_Auto_Ptr.cpp#L11-L15

Here the object is going away so it doesn't really matter either.

So it looks like it's not nulling the value just because it doesn't always strictly have to. I do admit it would be the safe thing to do though.

Do you see a use-after-free issue with this or is this just a question?

**Reply by ashaikh111** [reply]
Thanks Fred for the reply.
I am using ACE library in my one of the applications, which is crashing at client side. When I open crash dump in visual studio 2015 then it shows control at following statement:
if (rep->ref_count_-- == 0) {
So, I am trying to understand, why "0" or NULL is not assigned to it immediately in the same function I.e. in "Detach" and "0" is assigned out side this method. Why it is kept dangling? I am trying to understand purpose. If we don't assign "0" after free then, is it unsafe?
